presto-parquet/src/main/java/io/prestosql/parquet/writer/ParquetSchemaConverter.java

```diff
@@ -124,9 +128,27 @@ else if (decimalType.isShort()) {
         if (DATE.equals(type)) {
             return Types.optional(PrimitiveType.PrimitiveTypeName.INT32).as(OriginalType.DATE).named(name);
         }
-        if (BIGINT.equals(type) || TIMESTAMP.equals(type)) {
+        if (BIGINT.equals(type)) {
             return Types.primitive(PrimitiveType.PrimitiveTypeName.INT64, OPTIONAL).named(name);
         }
+
+        if (type instanceof TimestampType) {
+            TimestampType timestampType = (TimestampType) type;
+
+            LogicalTypeAnnotation.TimeUnit timeUnit;
+            if (timestampType.getPrecision() <= 3) {
+                timeUnit = LogicalTypeAnnotation.TimeUnit.MILLIS;
+            }
+            else if (timestampType.getPrecision() <= 6) {
+                timeUnit = LogicalTypeAnnotation.TimeUnit.MICROS;
+            }
+            else {
+                throw new PrestoException(NOT_SUPPORTED, "Parquet support not implemented for TIMESTAMP with precision: " + timestampType.getPrecision());
+            }
+
+            LogicalTypeAnnotation logicalTypeAnnotation = LogicalTypeAnnotation.timestampType(session.isLegacyTimestamp(), timeUnit);
```

Comment:
Presto timestamps are not supposed to be instants. The legacy timestamp semantics are an accident and should not be encoded into the logical type in a data file. We should instead be translating the legacy value at read and write.

Reply:
Ok. [The field](https://github.com/apache/parquet-format/blob/345282ce307a9c9dcc15b6f3c9106be379ec26ba/LogicalTypes.md#timestamp) seemed to have the same meaning as the legacy timestamp flag but I can remove it. If presto timestamps are not supposed to be instants should the logical type be hardcoded to match, or should it be configurable via session/catalog properties?

Reply:
The implementation and behavior is the same, but logical meaning is not. It’s also largely hidden from users, and something we hope to completely fix soon. We only allow writing in Hive if the session time zone matches the JVM time zone, which contributes to the difference being hidden. It’s a complicated topic.